apps/src/lib/config/genesis.rs

```diff
@@ -1,17 +1,426 @@
+    };
+
+    #[derive(Clone, Debug, Deserialize, Serialize)]
+    pub struct HexString(pub String);
+
+    impl HexString {
+        pub fn to_bytes(&self) -> Result<Vec<u8>, HexKeyError> {
+            let bytes = hex::decode(self.0.to_owned())?;
+            Ok(bytes)
+        }
+
+        pub fn to_sha256_bytes(&self) -> Result<[u8; 32], HexKeyError> {
+            let bytes = hex::decode(self.0.to_owned())?;
+            let slice = bytes.as_slice();
+            let array: [u8; 32] = slice.try_into()?;
+            Ok(array)
+        }
+
+        pub fn to_public_key(&self) -> Result<PublicKey, HexKeyError> {
+            let key = PublicKey::from_str(&self.0)?;
+            Ok(key)
+        }
+    }
+
+    #[derive(Debug)]
+    pub enum HexKeyError {
+        InvalidHexString(hex::FromHexError),
+        InvalidSha256(TryFromSliceError),
+        InvalidPublicKey(ParsePublicKeyError),
+    }
+
+    impl From<hex::FromHexError> for HexKeyError {
+        fn from(err: hex::FromHexError) -> Self {
+            Self::InvalidHexString(err)
+        }
+    }
+
+    impl From<ParsePublicKeyError> for HexKeyError {
+        fn from(err: ParsePublicKeyError) -> Self {
+            Self::InvalidPublicKey(err)
+        }
+    }
+
+    impl From<TryFromSliceError> for HexKeyError {
+        fn from(err: TryFromSliceError) -> Self {
+            Self::InvalidSha256(err)
+        }
+    }
+
+    #[derive(Clone, Debug, Deserialize, Serialize)]
+    pub struct GenesisConfig {
+        // Genesis timestamp
+        pub genesis_time: Rfc3339String,
+        // Initial validator set
+        pub validator: HashMap<String, ValidatorConfig>,
+        // Token accounts present at genesis
+        pub token: Option<HashMap<String, TokenAccountConfig>>,
+        // Established accounts present at genesis
+        pub established: Option<HashMap<String, EstablishedAccountConfig>>,
+        // Implicit accounts present at genesis
+        pub implicit: Option<HashMap<String, ImplicitAccountConfig>>,
+        // Protocol parameters
+        pub parameters: ParametersConfig,
+        // PoS parameters
+        pub pos_params: PosParamsConfig,
+        // Wasm definitions
+        pub wasm: HashMap<String, WasmConfig>,
+    }
+
+    #[derive(Clone, Debug, Deserialize, Serialize)]
+    pub struct ValidatorConfig {
+        // Public key for consensus. (default: generate)
+        pub consensus_public_key: Option<HexString>,
+        // Public key for validator account. (default: generate)
+        pub account_public_key: Option<HexString>,
+        // Public key for staking reward account. (default: generate)
+        pub staking_reward_public_key: Option<HexString>,
+        // Validator address (default: generate).
+        pub address: Option<String>,
+        // Staking reward account address (default: generate).
+        pub staking_reward_address: Option<String>,
+        // Total number of tokens held at genesis.
+        // XXX: u64 doesn't work with toml-rs!
```

Comment:
opened https://github.com/anoma/anoma/issues/525. I think maybe for tokens we should define a custom type, since we're using `token::Amount::whole` to construct it, which has a max value `9223372036854` (`(i64::MAX - 1) / 10^6`)